library/core/src/str/iter.rs

```diff
@@ -1439,11 +1439,20 @@ impl<'a> Iterator for EncodeUtf16<'a> {
 
     #[inline]
     fn size_hint(&self) -> (usize, Option<usize>) {
-        let (low, high) = self.chars.size_hint();
-        // every char gets either one u16 or two u16,
-        // so this iterator is between 1 or 2 times as
-        // long as the underlying iterator.
-        (low, high.and_then(|n| n.checked_mul(2)))
+        let len = self.chars.iter.len();
+        // The highest bytes:code units ratio occurs for 3-byte sequences, so
```

Comment:
I think this could explicitly mention surrogates, something like "because a 4-byte sequence will map to 2 code units as a surrogate pair," so we're covering the prior misconception about the output being twice as long.

Everything else looks good to me!

Reply:
Added some extra details.